src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs

```diff
@@ -0,0 +1,138 @@
+        }
+
+        public static unsafe SafeMsQuicConfigurationHandle Create(QuicListenerOptions options)
+        {
+            // TODO: lots of ServerAuthenticationOptions are not yet supported by MsQuic.
+            return Create(options, QUIC_CREDENTIAL_FLAGS.NONE, options.ServerAuthenticationOptions?.ServerCertificate, options.ServerAuthenticationOptions?.ApplicationProtocols);
+        }
+
+        // TODO: this is called from MsQuicListener and when it fails it wreaks havoc in MsQuicListener finalizer.
+        //       Consider moving bigger logic like this outside of constructor call chains.
+        private static unsafe SafeMsQuicConfigurationHandle Create(QuicOptions options, QUIC_CREDENTIAL_FLAGS flags, X509Certificate? certificate, List<SslApplicationProtocol>? alpnProtocols)
+        {
+            // TODO: some of these checks should be done by the QuicOptions type.
+            if (alpnProtocols == null || alpnProtocols.Count == 0)
+            {
+                throw new Exception("At least one SslApplicationProtocol value must be present in SslClientAuthenticationOptions or SslServerAuthenticationOptions.");
+            }
+
+            if (options.MaxBidirectionalStreams > ushort.MaxValue)
+            {
+                throw new Exception("MaxBidirectionalStreams overflow.");
+            }
+
+            if (options.MaxBidirectionalStreams > ushort.MaxValue)
+            {
+                throw new Exception("MaxBidirectionalStreams overflow.");
+            }
+
+            Debug.Assert(!MsQuicApi.Api.Registration.IsInvalid);
+
+            var settings = new QuicSettings
+            {
+                IsSetFlags = QuicSettingsIsSetFlags.PeerBidiStreamCount |
+                             QuicSettingsIsSetFlags.PeerUnidiStreamCount,
+                PeerBidiStreamCount = (ushort)options.MaxBidirectionalStreams,
+                PeerUnidiStreamCount = (ushort)options.MaxUnidirectionalStreams
+            };
+
+            if (options.IdleTimeout != Timeout.InfiniteTimeSpan)
+            {
+                if (options.IdleTimeout <= TimeSpan.Zero) throw new Exception("IdleTimeout must not be negative.");
+
+                ulong ms = (ulong)options.IdleTimeout.Ticks / TimeSpan.TicksPerMillisecond;
+                if (ms > (1ul << 62) - 1) throw new Exception("IdleTimeout is too large (max 2^62-1 milliseconds)");
+
+                settings.IsSetFlags |= QuicSettingsIsSetFlags.IdleTimeoutMs;
+                settings.IdleTimeoutMs = (ulong)options.IdleTimeout.TotalMilliseconds;
+            }
+
+            uint status;
+            SafeMsQuicConfigurationHandle? configurationHandle;
+
+            MemoryHandle[]? handles = null;
+            QuicBuffer[]? buffers = null;
+            try
+            {
+                MsQuicAlpnHelper.Prepare(alpnProtocols, out handles, out buffers);
```

Comment:
I don't think we need safe handles here.  Since this only needs preserve through the call and never be exposed outside of this function I'm wondering if there is easier way. I'm sure the handles will work but feel heavy.

Reply:
I see:
https://github.com/microsoft/msquic/blob/55c15aeb4f82335959fefa86005e69d1e7a81c58/src/core/configuration.c#L96-L105

But is this something we can safely assume will be true even in the future? Docs don't say anything about it: https://github.com/microsoft/msquic/blob/main/docs/api/ConfigurationOpen.md

cc: @nibanks

Reply:
Yes, you can rely on MsQuic to copy what it needs in the function call. The memory doesn't have to exist beyond that,

Reply:
Ultimately I'd love to have a configuration API -- something like `SocketsHttpHandler` really doesn't need more than a single configuration, while the current code will allocate one for every new connection it opens.

Reply:
> Ultimately I'd love to have a configuration API -- something like `SocketsHttpHandler` really doesn't need more than a single configuration, while the current code will allocate one for every new connection it opens.

Creating a configuration is a very heavy weight operation. I'd highly recommend one of the next things you do is to refactor this like @scalablecory mentions.

Reply:
Does that mean that `ConfigurationLoadCredential` might be called multiple times on the same configuration with different certificate?
Or how should I go about sharing the configuration?

Am I wrong in my train of thought: instance of `SocketsHttpHandler` corresponds to one connection pool which corresponds to a single `QuicConnection` which holds one configuration.

Reply:
Generally, you create one configuration for each unique security configuration. On server that usually maps 1:1 with each server certificate. On client most times there is only one, unless they're creating clients with different configurations.

Reply:
(This is captured in https://github.com/dotnet/runtime/issues/42642)